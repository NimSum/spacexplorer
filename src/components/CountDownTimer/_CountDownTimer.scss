section.countdown-timer {
  display: flex;
  position: relative;
  z-index: 1;
  width: 60%;
  margin: 30px auto;
  text-align: center;
  font-size: 4rem;


  p {
    flex: 3;
    margin: 0;
  }

  span {
    flex: 1;
  }
}