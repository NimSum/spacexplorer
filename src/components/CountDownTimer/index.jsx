import React, { Component } from 'react'
import PropTypes from 'prop-types';

export class CountDownTimer extends Component {

  constructor() {
    super();
    this.state = {
      days: 0,
      hours: 0,
      minutes: 0,
      seconds: 0,
      totalTime: 0
    };
    this.timer  = setInterval(() => {
        this.setState({ 
          totalTime: this.state.totalTime - 1000
          }, () => this.calculateCountdown(this.state.totalTime));
      },
      1000
    );
  }

  calculateCountdown() {
    let { days, hours, minutes, seconds, totalTime } = this.state;
    days = Math.floor(totalTime / 86400000);
    totalTime = totalTime % 8640000;
    hours = Math.floor(totalTime / 3600000);
    totalTime = totalTime % 3600000;
    minutes = Math.floor(totalTime / 60000); 
    totalTime = totalTime % 60000;
    seconds = Math.floor(totalTime/1000);
    this.setState({ 
      days: days < 0 ? 0 : days, 
      hours: hours < 0 ? 0 : hours, 
      minutes: minutes < 0 ? 0 : minutes, 
      seconds: seconds < 0 ? 0 :  seconds
    })
  }
  
  componentDidMount() {
    const totalTime = this.calculateDateDifference();
    this.setState({ totalTime });
  }

  calculateDateDifference() {
    const nextDate = new Date(this.props.date);
    return (nextDate.getTime() - Date.now());
  }

  componentDidUpdate(prevProps) {
    if(this.props.date !== prevProps.date) {
      const totalTime = this.calculateDateDifference();
      this.setState({ totalTime })
    }
  }

  componentWillUnmount() {
    clearInterval(this.timer);
  }

  render() {
    const { days, hours, minutes, seconds } = this.state;
    return (
      <section className="countdown-timer">
        <div>
          <p>{ days < 10 ? `0${days}` : days }</p>
          <span>Days</span>
        </div>
        <span className="colon">:</span>
        <div>
          <p>{ hours < 10 ? `0${hours}` : hours }</p>
          <span>Hours</span>
        </div>
        <span className="colon">:</span>
        <div>
          <p>{ minutes < 10 ? `0${minutes}`: minutes }</p>
          <span>Minutes</span>
        </div>
        <span className="colon">:</span>
        <div>
          <p>{ seconds < 10 ? `0${seconds}` : seconds }</p>
          <span>Seconds</span>
        </div>
      </section>
    )
  }
}

CountDownTimer.propTypes = {
  date: PropTypes.string
}

export default CountDownTimer
