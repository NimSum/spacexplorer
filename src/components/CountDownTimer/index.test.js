import React from 'react';
import { shallow } from 'enzyme';
import CountDownTimer from './index';


global.Date.getTime = jest.fn().mockImplementation(() => 10);
global.Date.now = jest.fn().mockImplementation(() => 1);

describe('CountDownTimer', () => {
  const mockDate = '20120-06-15T00:00:00Z';

  let spyCalcDifference;
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(
      < CountDownTimer 
        date={ mockDate } 
      />)
    spyCalcDifference = jest.spyOn(wrapper.instance(), 'calculateDateDifference')
  })

  it('should match component snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  })

  it('should should set state calculated difference on mount', () => {
    wrapper.instance().componentDidMount();
    expect(typeof wrapper.state().totalTime).toEqual('number');
    expect(spyCalcDifference).toHaveBeenCalledTimes(1)
  })

  it('should re-calculate difference with updated date prop', () => {
    const date = '2021-06-15T00:00:00Z';
    wrapper.setProps({ date });
    expect(spyCalcDifference).toHaveBeenCalledTimes(1);
    expect(typeof wrapper.state().totalTime).toEqual('number');
  })

  it.skip('should calculate date difference', () => {
    const result = wrapper.instance().calculateDateDifference();
    const expected = 9;
    expect(result).toEqual(expected);
  })

  it('should calculate countdown and set it to state', () => {
    wrapper.setState({ totalTime: 100000000 });
    const expected = { 
      days: 1,
      hours: 1,
      minutes: 22,
      seconds: 40,
      totalTime: 100000000 
    }
    wrapper.instance().calculateCountdown();
    expect(wrapper.state()).toEqual(expected);
    expect(wrapper).toMatchSnapshot();
  })

  it.skip('should update timer every second', () => {
    jest.useFakeTimers();
    wrapper.instance().updateTime();
    jest.runAllTimers();
    expect(setInterval).toHaveBeenCalledTimes(1);
    expect(setInterval).toHaveBeenLastCalledWith(expect.any(Function), 1000);
  })  

})