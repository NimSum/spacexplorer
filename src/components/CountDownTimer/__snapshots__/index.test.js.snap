// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountDownTimer should calculate countdown and set it to state 1`] = `
<section
  className="countdown-timer"
>
  <div>
    <p>
      01
    </p>
    <span>
      Days
    </span>
  </div>
  <span
    className="colon"
  >
    :
  </span>
  <div>
    <p>
      01
    </p>
    <span>
      Hours
    </span>
  </div>
  <span
    className="colon"
  >
    :
  </span>
  <div>
    <p>
      22
    </p>
    <span>
      Minutes
    </span>
  </div>
  <span
    className="colon"
  >
    :
  </span>
  <div>
    <p>
      40
    </p>
    <span>
      Seconds
    </span>
  </div>
</section>
`;

exports[`CountDownTimer should match component snapshot 1`] = `
<section
  className="countdown-timer"
>
  <div>
    <p>
      00
    </p>
    <span>
      Days
    </span>
  </div>
  <span
    className="colon"
  >
    :
  </span>
  <div>
    <p>
      00
    </p>
    <span>
      Hours
    </span>
  </div>
  <span
    className="colon"
  >
    :
  </span>
  <div>
    <p>
      00
    </p>
    <span>
      Minutes
    </span>
  </div>
  <span
    className="colon"
  >
    :
  </span>
  <div>
    <p>
      00
    </p>
    <span>
      Seconds
    </span>
  </div>
</section>
`;
